File: webflux-dialogue/src/main/java/com/study/webflux/rag/infrastructure/monitoring/config/TtsMetricsConfiguration.java
package com.study.webflux.rag.infrastructure.monitoring.config;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import com.study.webflux.rag.infrastructure.dialogue.adapter.tts.loadbalancer.TtsEndpoint;
import com.study.webflux.rag.infrastructure.dialogue.adapter.tts.loadbalancer.TtsLoadBalancer;
import io.micrometer.core.instrument.Gauge;
import io.micrometer.core.instrument.binder.MeterBinder;

/**
 * TTS 로드밸런서 메트릭을 Prometheus에 노출합니다.
 */
@Configuration
public class TtsMetricsConfiguration {

	@Bean
	public MeterBinder ttsEndpointMetrics(TtsLoadBalancer loadBalancer) {
		return registry -> {
			for (TtsEndpoint endpoint : loadBalancer.getEndpoints()) {
				Gauge
					.builder("tts.endpoint.active.requests",
						endpoint,
						TtsEndpoint::getActiveRequests)
					.description("TTS 엔드포인트별 활성 요청 수")
					.tag("endpoint", endpoint.getId())
					.register(registry);

				Gauge.builder("tts.endpoint.health", endpoint, ep -> healthToNumber(ep.getHealth()))
					.description("TTS 엔드포인트 상태 (1=HEALTHY, 0=TEMP_FAIL, -1=PERM_FAIL)")
					.tag("endpoint", endpoint.getId())
					.register(registry);
			}
		};
	}

	private double healthToNumber(TtsEndpoint.EndpointHealth health) {
		return switch (health) {
			case HEALTHY -> 1;
			case TEMPORARY_FAILURE -> 0;
			case PERMANENT_FAILURE -> -1;
			case CLIENT_ERROR -> 0;
		};
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/infrastructure/monitoring/config/TtsMetricsConfiguration.kt
package com.miyou.app.infrastructure.monitoring.config

import com.miyou.app.infrastructure.dialogue.adapter.tts.loadbalancer.TtsEndpoint
import com.miyou.app.infrastructure.dialogue.adapter.tts.loadbalancer.TtsEndpoint.EndpointHealth
import com.miyou.app.infrastructure.dialogue.adapter.tts.loadbalancer.TtsLoadBalancer
import io.micrometer.core.instrument.Gauge
import io.micrometer.core.instrument.binder.MeterBinder
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class TtsMetricsConfiguration {
    @Bean
    fun ttsEndpointMetrics(loadBalancer: TtsLoadBalancer): MeterBinder =
        MeterBinder { registry ->
            loadBalancer.endpoints.forEach { endpoint ->
                Gauge
                    .builder("tts.endpoint.active.requests", endpoint) { it.activeRequests.toDouble() }
                    .description("TTS endpoint active requests")
                    .tag("endpoint", endpoint.id)
                    .register(registry)

                Gauge
                    .builder("tts.endpoint.health", endpoint) { ep -> healthToNumber(ep.health) }
                    .description("TTS endpoint health status (1=HEALTHY, 0=TEMP_FAIL, -1=PERM_FAIL)")
                    .tag("endpoint", endpoint.id)
                    .register(registry)
            }
        }

    private fun healthToNumber(health: EndpointHealth): Double =
        when (health) {
            EndpointHealth.HEALTHY -> 1.0
            EndpointHealth.TEMPORARY_FAILURE -> 0.0
            EndpointHealth.PERMANENT_FAILURE -> -1.0
            EndpointHealth.CLIENT_ERROR -> 0.0
        }
}
